[DYN-2797] Provide custom node info in the IronPython2 migration dialog.

reddyashish opened the pull request:
### Purpose

This PR implements UI functionality, to provide more information on the custom nodes in the workspace that have a direct dependency on IronPython2. This info is shown categorically in the Python migration dialog so that the users will have a better understanding of what they need to update. 

The list is shown in a tree view to differentiate between CN's that are from packages and the ones that are user defined. 
If the CN is user defined, then users can double click on the CN item and switch to its workspace to update the custom node to use CPython3 engine. If the CN is from a specific package, we show the package name and would expect the package authors to update the packages to use CPython3. The items in the list can be copy/paste-able. 

Based on the mockup provided in the task (https://jira.autodesk.com/browse/DYN-2797), I have created a first look of how the dialog would look like with all this information. 

![iron python dialog 2](https://user-images.githubusercontent.com/43763136/86849461-b2b6f080-c07d-11ea-9c31-8258b5184de2.gif)

As of now, the custom nodes from a single package are not grouped together and are instead shown as separate list items. To achieve this, we might need another dictionary which would contain all the list of CN's per package id. Also binding wpf dictionary data with multiple values seems little complicated. 

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@DynamoDS/dynamo

Amoursol commented:
@reddyashish looking really good! It's super clean and will be a huge addition that helps users understand the change. 

A couple of thoughts:
- Do we have a way for them to notify the package author? I'm not sure if due to GDPR we are publishing any personal details.
- If you select a node in the **Loose** section, does it zoom to fit on that node in the graph?

reddyashish commented on src/Libraries/DSCPython/DSCPython.csproj:
```diff
@@ -25,6 +25,7 @@
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
     <Prefer32Bit>false</Prefer32Bit>
+    <LangVersion>7.1</LangVersion>
```
I had to update the lang version here to 7.1 to get the DSCPython project build. I was getting this error: https://github.com/DynamoDS/Dynamo/pull/10830#issuecomment-652030087. @mmisol How did you get around it?

reddyashish replied:
I was using 2017 VS. Will move to 2019 and remove this change.

reddyashish commented:
@Amoursol I am not aware of any way, we can let users notify the package author directly from Dynamo. So I just added a tooltip for now. @mjkkirschner any thoughts?
The custom nodes shown here are the nodes that has a python script inside its workspace directly. So it can be that the CN shown here is not present directly in the workspace the user has opened but could be inside some nested custom nodes. It would get complicated if we want to zoom to fit into the selected CN.

Amoursol commented:
@reddyashish - Just to confirm, there are two sections: One = **Packaged** and two = **Loose**. Is the Packaged one inside Custom Nodes, and the Loose one just inside the graph itself? If so - can we zoom to fit on the Loose nodes in graph, the same way we do from **TuneUp**?

reddyashish commented:
Packaged refers to the Custom nodes that are defined in packages and Loose refers to the CN's that are user defined and are stored in %AppData%\Roaming\Dynamo\Dynamo Core\2.8\definitions. So Loose CN's can be used a different custom node.

Amoursol commented:
@reddyashish Ah, I see! Yes, in that case users will not understand 😄  Can we please change the loose name to **Definitions**? This matches the %appdata% name.

reddyashish commented:
@Amoursol  Changed the name to Definitions. This is how it is now:
<img width="666" alt="Screen Shot 2020-07-09 at 11 00 47 AM" src="https://user-images.githubusercontent.com/43763136/87056414-95943600-c1d3-11ea-9374-3338d590d923.png">

Amoursol commented:
@reddyashish LGTM

mjkkirschner commented on src/PythonMigrationViewExtension/GraphPythonDependencies.cs:
```diff
@@ -14,9 +14,9 @@ public class GraphPythonDependencies
 
         // A dictionary to mark Custom Nodes if they have a IronPython dependency or not. 
-        private static Dictionary<Guid, CNPythonDependency> CustomNodePythonDependency = new Dictionary<Guid, CNPythonDependency>();
+        internal static Dictionary<Guid, CNPythonDependency> CustomNodePythonDependency = new Dictionary<Guid, CNPythonDependency>();
```
is this internal for testing or so it can be called from some other assembly?

mjkkirschner replied:
because this an internal field I guess it should be camelCase.

reddyashish replied:
Internal because it is being used by other files.

mjkkirschner commented on src/PythonMigrationViewExtension/GraphPythonDependencies.cs:
```diff
@@ -14,9 +14,9 @@ public class GraphPythonDependencies
 
         // A dictionary to mark Custom Nodes if they have a IronPython dependency or not. 
-        private static Dictionary<Guid, CNPythonDependency> CustomNodePythonDependency = new Dictionary<Guid, CNPythonDependency>();
+        internal static Dictionary<Guid, CNPythonDependency> CustomNodePythonDependency = new Dictionary<Guid, CNPythonDependency>();
```
to me this sounds like a singular item, maybe add Map to the name?

mjkkirschner commented on src/PythonMigrationViewExtension/GraphPythonDependencies.cs:
```diff
@@ -28,23 +28,29 @@ internal GraphPythonDependencies(ViewLoadedParams viewLoadedParams)
         }
 
-        internal bool ContainsIronPythonDependencies()
+        internal bool ContainsIronPythonDependency()
```
hmm, I think a more appropriate name would be `CurrentWorkspaceContainsIronPythonDependency` - what do you think?

reddyashish replied:
yes, sounds good. will change.

QilongTang replied:
Sorry about being picky on this,, https://stackoverflow.com/questions/1370840/naming-conventions-what-to-name-a-method-that-returns-a-boolean

Maybe `ContainsIronPythonDependencyInCurrentWS`

mjkkirschner commented on src/PythonMigrationViewExtension/GraphPythonDependencies.cs:
```diff
@@ -14,9 +14,9 @@ public class GraphPythonDependencies
+        internal static Dictionary<Guid, CNPythonDependency> CustomNodePythonDependency = new Dictionary<Guid, CNPythonDependency>();
 
-        private enum CNPythonDependency
+        internal enum CNPythonDependency
```
CNPythonDependencyType?

mjkkirschner commented on src/PythonMigrationViewExtension/IronPythonInfoDialog.xaml:
```diff
@@ -51,12 +50,57 @@
+                                            <ColumnDefinition Width="*" />
+                                        </Grid.ColumnDefinitions>
+                                        <TextBlock Foreground="#bbbbbb" Text="{Binding PackageName}" />
+                                        <TextBlock Grid.Column="1" Foreground="#bbbbbb" Text="{Binding Name}" />
```
can we use colors only from the dynamo resource dictionary and if we need new colors add them to the dictionary with clear names?

reddyashish replied:
Using the colors from resource dictionary now.

mjkkirschner commented on src/PythonMigrationViewExtension/IronPythonInfoDialog.xaml.cs:
```diff
@@ -12,8 +16,50 @@ internal IronPythonInfoDialog(PythonMigrationViewExtension viewModel)
+
+                    if (customNodeInfo.PackageInfo != null)
+                    {
+                        PackagedCustomNodesContainingPython.Items.Add(customNodeItem);
```
is there a reason to implement this manually instead of using a binding?

mjkkirschner replied:
also if you do decide to use a binding remember to use an observable collection.

reddyashish replied:
As this list is static and doesn't have much functionality added to it, we decided to keep it simple.

mjkkirschner commented on src/PythonMigrationViewExtension/IronPythonInfoDialog.xaml.cs:
```diff
@@ -12,8 +16,50 @@ internal IronPythonInfoDialog(PythonMigrationViewExtension viewModel)
+                    }
+                }
+
+                if (PackagedCustomNodesContainingPython.Items.Count > 0)
```
same question here?

mjkkirschner commented on src/PythonMigrationViewExtension/IronPythonInfoDialog.xaml.cs:
```diff
@@ -24,5 +70,56 @@ private void OnMoreInformationButtonClicked(object sender, RoutedEventArgs e)
+        {
+            if (UserDefinitionCustomNodesContainingPython.SelectedItems.Count == 1)
+            {
+                CustomNodeItem selectedCNItem = UserDefinitionCustomNodesContainingPython.SelectedItem as CustomNodeItem;
```
could this ever be null after this `as`?

reddyashish replied:
I dont think it can be null here, as we are checking that the only one list item is selected. and the UserDefinitionCustomNodesContainingPython contains only user defined custom nodes.

reddyashish replied:
Do you think the OpenCustomNodeWorkspace should be called from try-catch?

mjkkirschner commented on src/PythonMigrationViewExtension/IronPythonInfoDialog.xaml.cs:
```diff
@@ -24,5 +70,56 @@ private void OnMoreInformationButtonClicked(object sender, RoutedEventArgs e)
+
+        public override string ToString()
+        {
+            if (PackageName.Equals("Definitions"))
```
can we pull this into a shared constant?

mjkkirschner commented on src/PythonMigrationViewExtension/IronPythonInfoDialog.xaml:
```diff
@@ -51,12 +50,57 @@
+                            </ListBox.ItemTemplate>
+                        </ListBox>
+                    </Expander>
+                    <Expander Grid.Row="1" x:Name="ExpanderForUserDefinitionDependencies" Header="Definitions"  Foreground="#bbbbbb" FontSize="15" Margin="20,5,0,0" Visibility="Collapsed">
```
The word definitions needs to be added to the localization resx.

mjkkirschner commented on src/PythonMigrationViewExtension/IronPythonInfoDialog.xaml:
```diff
@@ -51,12 +50,57 @@
+            <Expander x:Name="MainExpander" Header="{x:Static w:Resources.CustomNodesPythonDependencyHeader}" Foreground="#bbbbbb"
+                      FontSize="15" Margin="-2,0,0,0" Visibility="Hidden" ScrollViewer.VerticalScrollBarVisibility="Visible">
+                <StackPanel Height="Auto">
+                    <Expander x:Name="ExpanderForPackagedDependencies" Header="Packaged" Foreground="#bbbbbb" FontSize="15" Margin="20,5,0,0" Visibility="Collapsed">
```
"Packaged" needs to be added to the localization resx.

mjkkirschner commented:
@reddyashish some comments - also tests 😉 ?

reddyashish commented:
I have added tests for CustomNodesContainIronPythonDependency API in the previous PR, I thought we might need UI automation tests for this? What do you think?

reddyashish commented:
I think I addressed all the comments and would be good to test this as well on Monday's bug bash. So If this looks, we can merge it and will address any issues in a followup PR.

mjkkirschner commented on src/PythonMigrationViewExtension/IronPythonInfoDialog.xaml.cs:
```diff
@@ -24,5 +67,65 @@ private void OnMoreInformationButtonClicked(object sender, RoutedEventArgs e)
+            
+            foreach (var item in lb.SelectedItems)
+            {
+                collectedText += item.ToString() + "\r\n";
```
just a small thing, but I think this should be `System.Environment.NewLine`

reddyashish commented:
Thanks @mjkkirschner and @QilongTang. Merging this.
